system/core/tags.php

```diff
@@ -0,0 +1,184 @@
+	}
+	function addTag() {
+		var name = prompt('<?=__('enter-label')?>', '');
+		if (name) {
+			var description = prompt('<?=__('enter-description')?>', '');
+			if (description) {
+				var json = '{ "name" : "' + name + '", "description" : "' + description + '"}';
+				hypha('<?=$url?>', 'tagNew', json, function(response) {document.getElementById('tagList').innerHTML = response;});
+				document.getElementById('tagList').innerHTML = 'loading...';
+			}
+		}
+	}
+</script>
+<?php
+		$hyphaHtml->writeScript(ob_get_clean());
+		
+		return tagList($pageListNode, $lang);
+	}
+	
+	registerCommandCallback('tagSelect', 'tagSelectCallback');
+	registerCommandCallback('tagDeselect', 'tagDeselectCallback');
+	registerCommandCallback('tagNew', 'tagNewCallback');
+	
+	function tagSelectCallback($id) {
+		global $hyphaXml;
+		global $O_O;
+		
+		$lang = $O_O->getContentLanguage();
+		
+		$hyphaXml->lockAndReload();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+		$newtag = $hyphaXml->createElement('tag');
+		$newtag->setAttribute('id', $id);
+		$pageListNode->appendChild($newtag);
+		$hyphaXml->saveAndUnlock();
+		
+		echo tagList($pageListNode, $lang);
+		exit;
+	}
+	
+	function tagDeselectCallback($id) {
+		global $hyphaXml;
+		global $O_O;
+		$lang = $O_O->getContentLanguage();
+		
+		$hyphaXml->lockAndReload();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+		foreach($pageListNode->getElementsByTagName('tag') as $tag) if ($tag->getAttribute('id')==$id) $pageListNode->removeChild($tag);
+		$hyphaXml->saveAndUnlock();
+		
+		echo tagList($pageListNode, $lang);
+		exit;
+	}
+	
+	function tagNewCallback($json) {
+		global $hyphaXml;
+		global $O_O;
+		
+		$lang = $O_O->getContentLanguage();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+		
+		$json = json_decode($json, true);
+		$newTagName = $json['name'];
+		$newTagDescription = $json['description'];
+		$newTagId = uniqid();
+		
+		// Check if tagList exists
+		if(!$hyphaXml->getElementsByTagName('tagList')->length) {
+			$hyphaXml->lockAndReload();
+			$tagList = $hyphaXml->createElement('tagList');
+			$hyphaXml->documentElement->appendChild($tagList);
+			$hyphaXml->saveAndUnlock();
+		}
+		else $tagList = $hyphaXml->getElementsByTagName('tagList')->Item(0);
+		
+		// Check if tag exists
+		$tagExists = false;
+		foreach($tagList->getElementsByTagName('tag') as $tag) {
+			foreach($tag->getElementsByTagName('language') as $language) {
+				if($language->getAttribute('id')==$lang && $language->getAttribute('label')==$json['name']) $tagExists = true;
+			}
+		}
+		
+		// Create new tag, add to tagList and page in hypha.xml
+		if (!$tagExists) {
+			$hyphaXml->lockAndReload();
+			$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+			
+			$tagList = $hyphaXml->getElementsByTagName('tagList')->Item(0);
+			$newTag = $hyphaXml->createElement('tag');
+			$newTag->setAttribute('id', $newTagId);
```

Comment:
This should use `$newTag->generateId()` (defined in database.php) rather than generating an id manually. This also means to use the `xml:id` attribute rather than `id` (which allows using getElementsById, IIRC).

Reply:
Fixed.